controllers/component_customizer.go

```diff
@@ -0,0 +1,328 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/kubernetes/scheme"
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+	"k8s.io/utils/pointer"
+	operatorv1 "sigs.k8s.io/cluster-api-operator/api/v1alpha1"
+	"sigs.k8s.io/cluster-api-operator/controllers/genericprovider"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+const (
+	deploymentKind       = "Deployment"
+	daemonSetKind        = "DaemonSet"
+	namespaceKind        = "Namespace"
+	managerContainerName = "manager"
+	// specHashAnnotation is added automatically to Deployments and DaemonSets.
+	// the value is the hash of the Spec. This is done so that if the Spec is updated
+	// then the annotation will also be updated, causing the pods to be rotated.
+	specHashAnnotation = "cluster.x-k8s.io/spec-hash"
```

Comment:
The deployment and daemonset controllers do this by default, why do we need to do this as well?